### PR TITLE
Ensuring reliability in asynс programming

### PR DESCRIPTION
**Pull request by LouiseMedova**

### CODE REVIEW

File: docs/developing-contracts/distributed_transactions.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+**Prepare phase:**
+The same steps of two phase commit protocol are followed here:
+- `Coordinator`:  
+The coordinator sends a prepare message to all participants and waits for replies;
+- `Participants`:  
+If the participants are ready to commit a transaction they send the ready message, otherwise they send no message to the coordinator. 
+- `Coordinator`:  
+Based on replies the coordinator decides either to go to the next state or not. If any of the participants respond with no message or if any of the participants fails to respond within a defined time, the coordinator sends an abort message to every participant.  It is important to highlight the differences from two phase commit protocol:
+   - The coordinator limits the response time from the participant. We can implement by sending a message with an indicated amount of gas or indicated the number of blocks the coordinator is ready to wait;
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:111]:
```suggestion
   - The coordinator limits the response time from the participant. We can implement this by sending a message with an indicated amount of gas or indicated number of blocks the coordinator is ready to wait;
```

File: docs/developing-contracts/distributed_transactions.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+2. Every phase can now time out – there is no indefinite wait as in the two-phase commit protocol.
+   - `Phase 1`:
+Participant aborts if it doesn’t hear from a coordinator in time;
+Coordinator sends aborts to all if it doesn’t hear from any participant.
+   - `Phase 2`:
+If a participant times out waiting for a coordinator, elect a new coordinator.
+
+Let’s get back to our swap contract.
+
+**Prepare phase**:
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:138]:
```suggestion
**Preparation phase**:  
```

File: docs/developing-contracts/distributed_transactions.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+**Prepare phase**:  
+The following cases are possible:
+- all token contracts receive the message;
+- the swap contract fails to wait for response from any token contract
+- the swap contract fails itself.
+
+In the case of a fall, if a transaction isn't restarted, the swap contract will not move to the second phase and the token contracts will unlock their state using delayed messages.
+![img alt](./img/3pc-1.png)
+**Pre-Commit phase**:
+At this stage we can have a failure in the swap contract or in the token contract only due the lack of gas.  To solve this problem we can use gas reservation as follows:
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:147]:
```suggestion
At this stage we can have a failure in the swap contract or in the token contract only due to the lack of gas.  To solve this problem we can use gas reservation as follows:
```

File: docs/developing-contracts/distributed_transactions.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+- all token contracts receive the message;
+- the swap contract fails to wait for response from any token contract
+- the swap contract fails itself.
+
+In the case of a fall, if a transaction isn't restarted, the swap contract will not move to the second phase and the token contracts will unlock their state using delayed messages.
+![img alt](./img/3pc-1.png)
+**Pre-Commit phase**:
+At this stage we can have a failure in the swap contract or in the token contract only due the lack of gas.  To solve this problem we can use gas reservation as follows: 
+- The swap contract receives the information about error in its `handle_signal`;
+- Using gas reservation (so, it’s necessary to care about gas reservations before), the swap contract sends a message to itself to restart the transaction from the second phase. (The same logic can be also used in `prepare phase`).
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:149]:
```suggestion
- Using gas reservation (so, it’s necessary to care about gas reservations before), the swap contract sends a message to itself to restart the transaction from the second phase. (The same logic can also be used in the `preparation phase`).
```

File: docs/developing-contracts/distributed_transactions.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+**Commit phase**:  
+As in the previous stage we can have a failure only due to the lack of gas. Here it is not so critical, since at this stage all participants can commit themselves.
+
+## Saga pattern
+**Theory**:  
+A `saga` is a sequence of local transactions. Each local transaction updates the database and publishes a message or event to trigger the next local transaction in the saga. If a local transaction fails because it violates a business rule then the saga executes a series of compensating transactions that undo the changes that were made by the preceding local transactions. Thus, Saga consists of multiple steps whereas `2PC` acts like a single request.  
+There are two ways of coordination sagas:
+- `Choreography` - each local transaction publishes domain events that trigger local transactions in other services;
+- `Orchestration` - an orchestrator (object) tells the participants what local transactions to execute.
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:161]:
```suggestion
- `Orchestration` - an orchestrator (object) tells the participants what local transactions to execute.
```

File: docs/developing-contracts/distributed_transactions.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+As in the previous stage we can have a failure only due to the lack of gas. Here it is not so critical, since at this stage all participants can commit themselves.
+
+## Saga pattern
+**Theory**:  
+A `saga` is a sequence of local transactions. Each local transaction updates the database and publishes a message or event to trigger the next local transaction in the saga. If a local transaction fails because it violates a business rule then the saga executes a series of compensating transactions that undo the changes that were made by the preceding local transactions. Thus, Saga consists of multiple steps whereas `2PC` acts like a single request.  
+There are two ways of coordination sagas:
+- `Choreography` - each local transaction publishes domain events that trigger local transactions in other services;
+- `Orchestration` - an orchestrator (object) tells the participants what local transactions to execute.  
+
+We will consider the `orchestration based Saga` where there would be an orchestrator(swap contract) to manage the entire operation from one center.
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:163]:
```suggestion
We will consider the `orchestration based Saga` where there would be an orchestrator (swap contract) to manage the entire operation from one center. 
```

File: docs/developing-contracts/distributed_transactions.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+## Saga pattern
+**Theory**:  
+A `saga` is a sequence of local transactions. Each local transaction updates the database and publishes a message or event to trigger the next local transaction in the saga. If a local transaction fails because it violates a business rule then the saga executes a series of compensating transactions that undo the changes that were made by the preceding local transactions. Thus, Saga consists of multiple steps whereas `2PC` acts like a single request.  
+There are two ways of coordination sagas:
+- `Choreography` - each local transaction publishes domain events that trigger local transactions in other services;
+- `Orchestration` - an orchestrator (object) tells the participants what local transactions to execute.  
+
+We will consider the `orchestration based Saga` where there would be an orchestrator(swap contract) to manage the entire operation from one center. 
+
+The swap operation consist of the following steps:
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:165]:
```suggestion
The swap operation consists of the following steps:
```

File: docs/developing-contracts/distributed_transactions.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+1. Swap contract receives a message to exchange tokens in the liquidity pool. So, it must transfer tokens A from the account to its address and then transfer tokens B to the user.
+2. It creates the first task: transfer tokens from user to swap contract. It also creates a compensating transaction for the first task: transfer tokens from the swap contract back to the user. The second task is to transfer tokens from the swap contract to the user.
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:167]:
```suggestion
2. It creates the first task: transfer tokens from the user to the swap contract. It also creates a compensating transaction for the first task: transfer tokens from the swap contract back to the user. The second task is to transfer tokens from the swap contract to the user.
```

File: docs/developing-contracts/distributed_transactions.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+3. It starts executing the first task. If the execution fails, it cancels the transaction. If it’s successful, the swap contract executes the second task;
+4. If the execution of the second task is successful, the transaction is completed. Otherwise, the swap contract executes the compensation transaction for the first task.
+![img alt](./img/saga.png)
+
+It is important to note that compensatory transactions should not fail due to any logical error. They can only fall due to lack of gas. If this happens, then you need to restart the transaction again or use the gas resevration. The `idempotency` of the token contract guarantees that the transaction will be completed to the end without any duplicate transactions.
```
Comment (shamilsan) [inline docs/developing-contracts/distributed_transactions.md:172]:
```suggestion
It is important to note that compensatory transactions should not fail due to any logical error. They can only fall due to lack of gas. If this happens, then you need to restart the transaction again or use the gas reservation. The `idempotency` of the token contract guarantees that the transaction will be completed to the end without any duplicate transactions.
```

**Review by shamilsan — APPROVED** [review]
The best technical article in our wiki! 👍

**Review by EugenWay — APPROVED** [review]

**Pull request merged**